File: ReactAndroid/src/main/java/com/facebook/react/animated/AnimatedNode.java
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * <p>This source code is licensed under the MIT license found in the LICENSE file in the root
 * directory of this source tree.
 */
package com.facebook.react.animated;

import androidx.annotation.Nullable;
import com.facebook.infer.annotation.Assertions;
import java.util.ArrayList;
import java.util.List;

/** Base class for all Animated.js library node types that can be created on the "native" side. */
/*package*/ abstract class AnimatedNode {

  public static final int INITIAL_BFS_COLOR = 0;

  private static final int DEFAULT_ANIMATED_NODE_CHILD_COUNT = 1;

  /*package*/ @Nullable List<AnimatedNode> mChildren; /* lazy-initialized when a child is added */
  /*package*/ int mActiveIncomingNodes = 0;
  /*package*/ int mBFSColor = INITIAL_BFS_COLOR;
  /*package*/ int mTag = -1;

  public final void addChild(AnimatedNode child) {
    if (mChildren == null) {
      mChildren = new ArrayList<>(DEFAULT_ANIMATED_NODE_CHILD_COUNT);
    }
    Assertions.assertNotNull(mChildren).add(child);
    child.onAttachedToNode(this);
  }

  public final void removeChild(AnimatedNode child) {
    if (mChildren == null) {
      return;
    }
    child.onDetachedFromNode(this);
    mChildren.remove(child);
  }

  /**
   * Subclasses may want to override this method in order to store a reference to the parent of a
   * given node that can then be used to calculate current node's value in {@link #update}. In that
   * case it is important to also override {@link #onDetachedFromNode} to clear that reference once
   * current node gets detached.
   */
  public void onAttachedToNode(AnimatedNode parent) {}

  /** See {@link #onAttachedToNode} */
  public void onDetachedFromNode(AnimatedNode parent) {}

  /**
   * This method will be run on each node at most once every repetition of the animation loop. It
   * will be executed on a node only when all the node's parent has already been updated. Therefore
   * it can be used to calculate node's value.
   */
  public void update() {}
}

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/animated/AnimatedNode.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.animated

import java.util.ArrayList

/** Base class for all Animated.js library node types that can be created on the "native" side. */
public abstract class AnimatedNode {

  internal companion object {
    internal const val INITIAL_BFS_COLOR: Int = 0
    internal const val DEFAULT_ANIMATED_NODE_CHILD_COUNT: Int = 1
  }

  // TODO: T196787278 Reduce the visibility of these fields to package once we have
  // converted the whole module to Kotlin

  @JvmField
  internal var children: MutableList<AnimatedNode>? =
      null /* lazy-initialized when a child is added */
  @JvmField internal var activeIncomingNodes: Int = 0
  @JvmField internal var BFSColor: Int = INITIAL_BFS_COLOR
  @JvmField internal var tag: Int = -1

  internal fun addChild(child: AnimatedNode) {
    val currentChildren =
        children
            ?: ArrayList<AnimatedNode>(DEFAULT_ANIMATED_NODE_CHILD_COUNT).also { children = it }

    currentChildren.add(child)
    child.onAttachedToNode(this)
  }

  internal fun removeChild(child: AnimatedNode): Unit {
    val currentChildren = children ?: return
    child.onDetachedFromNode(this)
    currentChildren.remove(child)
  }

  /**
   * Subclasses may want to override this method in order to store a reference to the parent of a
   * given node that can then be used to calculate current node's value in [update]. In that case it
   * is important to also override [onDetachedFromNode] to clear that reference once current node
   * gets detached.
   */
  internal open fun onAttachedToNode(parent: AnimatedNode): Unit = Unit

  /** See [onAttachedToNode] */
  internal open fun onDetachedFromNode(parent: AnimatedNode): Unit = Unit

  /**
   * This method will be run on each node at most once every repetition of the animation loop. It
   * will be executed on a node only when all the node's parent has already been updated. Therefore
   * it can be used to calculate node's value.
   */
  internal open fun update(): Unit = Unit

  /**
   * Pretty-printer for the AnimatedNode. Only called in production pre-crash for debug diagnostics.
   */
  internal abstract fun prettyPrint(): String

  internal fun prettyPrintWithChildren(): String {

    val currentChildren = children?.joinToString(" ")
    return prettyPrint() +
        if (!currentChildren.isNullOrBlank()) " children: $currentChildren" else ""
  }
}
